### PR TITLE
Move key changes into FormulaRuntime.

### PR DESCRIPTION
**Pull request by Laimiux**
## What
To remove duplication within RxJavaRuntime and FlowRuntime, I'm moving logic into shared `FormulaRuntime` class

### CODE REVIEW

**Comment by carrotkite** [comment]
### JaCoCo Code Coverage 78.73% :white_check_mark:
| Class | Covered | Meta | Status |
|:---|:---:|:---:|:---:|
| `com/instacart/formula/coroutines/FlowRuntime` | 50% | 0% | :white_check_mark: |
| `com/instacart/formula/rxjava3/RxJavaRuntime` | 84% | 0% | :white_check_mark: |
| `com/instacart/formula/FormulaRuntime` | 77% | 0% | :white_check_mark: |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: formula/src/main/java/com/instacart/formula/FormulaRuntime.kt [hunk 1]
```diff
@@ -15,11 +15,15 @@ class FormulaRuntime<Input : Any, Output : Any>(
     private val onOutput: (Output) -> Unit,
     private val onError: (Throwable) -> Unit,
     private val isValidationEnabled: Boolean = false,
-    private val inspector: Inspector? = null,
+    inspector: Inspector? = null,
 ) : ManagerDelegate {
     private val implementation = formula.implementation()
     private var manager: FormulaManagerImpl<Input, *, Output>? = null
-    private var hasInitialFinished = false
```
Comment (Laimiux) [inline formula/src/main/java/com/instacart/formula/FormulaRuntime.kt:22]: This is not needed anymore - previously it was used to ensure that on formula start, we wait for all actions to start running before emitting a value. Currently, it's done within `FormulaManagerImpl`.

File: formula-coroutines/src/main/java/com/instacart/formula/coroutines/FlowRuntime.kt [hunk 2]
```diff
@@ -27,35 +27,18 @@ object FlowRuntime {
...
-                )
-            }
-
-            var runtime = runtimeFactory()
-
-            input.onEach { input ->
-                threadChecker.check("Input arrived on a wrong thread.")
-                if (!runtime.isKeyValid(input)) {
-                    runtime.terminate()
-                    runtime = runtimeFactory()
```
Comment (Laimiux) [inline formula-coroutines/src/main/java/com/instacart/formula/coroutines/FlowRuntime.kt:52]: It was a bit weird that we recreate the runtime if key changes. Instead, the runtime will handle state reset internally.

**Review by jasonostrander — APPROVED** [review]

File: formula/src/main/java/com/instacart/formula/FormulaRuntime.kt [hunk 3]
```diff
@@ -135,7 +161,13 @@ class FormulaRuntime<Input : Any, Output : Any>(
                          */
                         if (manager.isTerminated()) {
                             shouldRun = false
-                            performTermination()
+                            terminateManager(manager)
+
+                            if (localInputId != inputId && !terminated) {
```
[the remark this answers is not in the document]
Comment (Laimiux) [reply formula/src/main/java/com/instacart/formula/FormulaRuntime.kt:166]: They could potentially both be true at the same time, will add documentation and maybe rename it.

**Pull request merged**